doc/basics.md

```diff
@@ -1066,6 +1086,91 @@ int main(void) {
 }
 ```
 
+### 2. Use `tag_invoke` for custom types
+
+Suppose you are writing this type:
+
+```C++
+/**
+ * A custom type that we want to parse.
+ */
+struct Car {
+  std::string make;
+  std::string model;
+
+  // tag_invoke is a friend
+  // the type of val will be a reference to one of these:
+  //     simdjson::ondemand::document
+  //     simdjson::ondemand::value
+  friend simdjson_result<Car>
+  tag_invoke(simdjson::deserialize_tag, std::type_identity<Car>, auto &val) {
+    simdjson::ondemand::object obj;
+    if (auto error = val.get_object().get(obj); error) {
+      return error;
+    }
+    Car car{};
+    // Instead of repeatedly obj["something"], we iterate through the object
+    // which we expect to be faster.
+    for (auto field : obj) {
+      simdjson::ondemand::raw_json_string key;
+      error = field.key().get(key);
+      if (error) {
+        return error;
+      }
+      if (key == "make") {
+        error = field.value().get_string(car.make);
+        if (error) {
+          return error;
+        }
+      } else if (key == "model") {
+        error = field.value().get_string(car.model);
+        if (error) {
+          return error;
+        }
+      }
+    }
+    return car;
+  }
+  
+};
+```
+
+Let us explain each argument of `tag_invoke`:
+
+- `simdjson::deserialize_tag`: it is the tag for Customization Point Object (CPO)
+- `std::type_identity<T>`: We specify the custom type we want here
+- `simdjson::ondemand::value&` or `simdjson::ondemand::document&`: the value or document that we want to deserialize from; you may want to just specify `auto&` to capture both of them.
+
+Suppose your custom types are `std::unique_ptr<any type>`, you could:
+
+```C++
+namespace simdjson {
```

Comment:
@lemire once C++20 (and maybe even reflecton in C++26) become heavily used. Do you think we should add these powerful generic tag_invoke definitions? Or that should always be up to the user?

We could provide:
- Support for PushableContainers<T> (https://github.com/simdjson/experimental_json_builder/blob/using_reflection_for_simdjson_ondemand/examples/example3.cpp)
- Support for unique_ptr as provided in this PR
- And even support of user defined types (:

Reply:
@the-moisrex @FranciscoThiesen Actually, I was thinking the same. My main concern is that we could be conservative because once we define something in the simdjson library, we make it difficult for users to do something different. So I don't know about "PushableContainers".  That seems too general.

Reply:
Hm, what about having a out-of-the-box specialization for user defined structs that use basic types?

Reply:
> Hm, what about having a out-of-the-box specialization for user defined structs that use basic types?

I have no strong opinion.